### PR TITLE
[SYCLomatic] Enable migration of __match_all/any_sync function

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/runtime/dpct-rt/include/util.hpp [hunk 1]
```diff
@@ -286,6 +286,73 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+unsigned int match_any_over_sub_group(sycl::sub_group g, unsigned member_mask,
+                                      T value) {
+  if (!member_mask) {
+    return 0;
+  }
+  unsigned int id = g.get_local_linear_id();
+  unsigned int flag = 0, result = 0, reduce_result = 0;
+  unsigned int bit_index = 0x1 << id;
+  bool is_participate = member_mask & bit_index;
+  T broadcast_value = 0;
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/util.hpp:311]: Add static assert `std::is_arithmetic`, or this line may cause unexpected error.

**Review by ziranzha — APPROVED** [review]

File: clang/runtime/dpct-rt/include/util.hpp [hunk 2]
```diff
@@ -286,6 +286,73 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+
+/// match_all_over_sub_group return mask and predicate \p pred will be set
+/// to 1 if all values that provided by each work-item are equal, otherwise
+/// return 0 and the predicate \p pred will be set to 0. The parameter
+/// \p member_mask indicating the work-items participating the call.
+/// \tparam T Input value type
+/// \param [in] g Input sub_group
+/// \param [in] member_mask Input mask
+/// \param [in] value Input value
+/// \param [in] pred Input predicate
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/util.hpp]:
```suggestion
/// \param [out] pred Input predicate
```

File: clang/runtime/dpct-rt/include/util.hpp [hunk 3]
```diff
@@ -286,6 +286,73 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
                                      : id);
 }
 
+/// match_any_over_sub_group return a mask in which some bits are set to 1,
+/// indicating that the values provided by the work-item represented by
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/util.hpp]: let's link the values to parameter value.

File: clang/runtime/dpct-rt/include/util.hpp [hunk 4]
```diff
@@ -286,6 +286,73 @@ T permute_sub_group_by_xor(sycl::sub_group g, T x, unsigned int mask,
...
+        g, is_participate ? (broadcast_value == value ? bit_index : 0) : 0,
+        sycl::plus<>());
+    flag |= reduce_result;
+    matched = reduce_result & bit_index;
+    result = matched * reduce_result + (1 - matched) * result;
+  }
+  return result;
+}
+
+/// match_all_over_sub_group return mask and predicate \p pred will be set
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/util.hpp]: link mask to member_mask

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**